client/lib/after-layout-flush/index.ts

```diff
@@ -11,19 +15,22 @@
  * Inspired by the Firefox performance best practices MDN article at:
  * https://developer.mozilla.org/en-US/Firefox/Performance_best_practices_for_Firefox_fe_engineers
  *
- * @param {Function} func The function to be invoked after the layout flush
- * @returns {Function} Returns the new delayed function
+ * @param func - The function to be invoked after the layout flush
+ * @returns The new delayed function
  */
-export default function afterLayoutFlush( func ) {
-	let rafHandle = undefined;
-	let timeoutHandle = undefined;
+export default function afterLayoutFlush< T extends ( ...args: any[] ) => any >( func: T ) {
```

Comment:
this _may_ be the equivalent of `<T extends Function>`

Reply:
I thought so too, but switching to that produces the following error further down in the `scheduleRAF` and `scheduleTimeout` `return`s:

```
Conversion of type '(this: any, ...args: any[]) => void' to type 'T' may be a mistake because neither type sufficiently overlaps with the other. If this was intentional, convert the expression to 'unknown' first.
```

Reply:
sorry for the delay - it's something we'd have to coordinate below.

also, can you help me understand the purpose of `lastThis` and the like? what are those there fore and why are we using them? are we having trouble with `this` binding and maintaining context?

Reply:
`afterLayoutFlush` is meant to be as bullet-proof as possible with regards to `this` usage, similar to `defer` and friends in Lodash. Because of that, we store and restore the `this` we receive, so that whatever the context the function is used in, everything will still work.